### PR TITLE
Backlog items: Copy ini files from primary to secondary

### LINKED ISSUE #376: Copying ini files from Primary to Secondary
## Problem Statement:
HSR setup ideally would be a role that can be used by itself without dependency.
Therefore there could be a case where one use this role to setup HSR from an existing HANA instance where ini file has been changed.

## Enhancement:
Add task to sync (one time) ini file configuration before setting up HSR.

Source: https://help.sap.com/viewer/6b94445c94ae495c83a19646e7c3fd56/2.0.02/en-US/86267e1ed56940bb8e4a45557cee0e43.html

### PR DESCRIPTION
**Pull request by robfallows**
## Problem

Closes #376

## Description

This is a lot more than copying `ini` files. The [SAP HANA Replication docs](https://help.sap.com/viewer/6b94445c94ae495c83a19646e7c3fd56/2.0.02/en-US/86267e1ed56940bb8e4a45557cee0e43.html) hint at this:

> The .ini file configuration must be similar for both systems...

Note the term "similar": i.e. not identical. It's not correct to simply copy `ini` files from the primary node to secondaries, as there is often host-specific configuration in those files.

The solution is to follow the recommendations in the [Monitoring and Replicating INI File Parameter Changes doc](https://help.sap.com/viewer/6b94445c94ae495c83a19646e7c3fd56/2.0.04/en-US/f65a9c62f60f4d4bb91620dd959ee68d.html) by setting the appropriate configuration in `global.ini` and letting HSR take care of replicating required settings.

## Pre-Requisites

1. Ensure you are on the correct branch: `git checkout centiq-backlog-items-copy-global-ini`
1. Ensure the branch is up to date: `git pull`

## Commitment to Test

- Full testing requires building and provisioning a two-node HA cluster and takes 1-2 hours - maybe longer if you hit Azure demand issues! Alternatively, you may review the pre-canned Ansible output at the bottom of these instructions.
- Reviewing new/changed files.

## Test Instructions / Acceptance Criteria

- [ ] Code changes are acceptable.
- [ ] Review the Ansible output for the `hana-system-replication` role, below for completeness.

### Build an HA Cluster

1. Follow the [usage document](https://github.com/Azure/sap-hana/blob/master/deploy/v2/USAGE.md), specifying `clustered_hana` as the template.
   - [ ] Infrastructure built successfully.
1. Log in to the RTI (the IP address is reported at the end of the infrastructure build).
1. Navigate to the `sap-hana` root and checkout the correct branch and ensure up-to-date:

   ```text
   cd sap-hana
   git checkout centiq-backlog-items-copy-global-ini
   git pull
   ```

1. Manually reconfigure the DB hosts to avoid [spontaneous hostname changes to `linux`](https://github.com/Azure/sap-hana/issues/346). :hand: run commands separately:

   ```text
   ansible hanadbnodes -i ~/hosts -b -a "sed -i 's/DHCLIENT_SET_HOSTNAME=.*/DHCLIENT_SET_HOSTNAME=\"no\"/' /etc/sysconfig/network/dhcp"
   ssh 10.1.1.4 "sudo hostname hdb1-0"
   ssh 10.1.1.5 "sudo hostname hdb1-1"
   ```

1. Navigate to the repo's `ansible` folder and execute the `sap_playbook.yml`:

   ```text
   cd deploy/v2/ansible
   ansible-playbook -i ~/hosts sap_playbook.yml
   ```

   - [ ] Playbook completes with no errors for `hdb1-0` and `hdb1-1`. :hand: You may see errors with other VMs.

<details>
<summary>Ansible Output (from start of `hana-system-replication` role)</summary>

```text
TASK [hana-system-replication : Check HANA system replication variables are set and seem valid] ********************************************************************************************
ok: [10.1.1.4] => {
    "changed": false,
    "msg": "All assertions passed"
}
ok: [10.1.1.5] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [hana-system-replication : Check that HANA was installed successfully and is running normally] ****************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Check whether replication has already been set up] *********************************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure current replication status is known] ****************************************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Check that HANA DB SYSTEM user can access the SYSTEM database] *********************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Check that HANA DB SYSTEM user can access the tenant database HN1] *****************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Check HANA log mode is set to normal] **********************************************************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure there is a local copy of Primary node SSFS_HN1.DAT file] ********************************************************************************************
skipping: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Ensure there is a local copy of Primary node SSFS_HN1.KEY file] ********************************************************************************************
skipping: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Check SSFS_HN1.DAT backup file exists on the Secondary node] ***********************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure SSFS_HN1.DAT file is backed up on the Secondary node] ***********************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Check SSFS_HN1.KEY backup file exists on the Secondary node] ***********************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure SSFS_HN1.KEY file is backed up on the Secondary node] ***********************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Ensure original SSFS_HN1.DAT file is removed on the Secondary node] ****************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Ensure original SSFS_HN1.KEY file is removed on the Secondary node] ****************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Ensure the Primary node SSFS_HN1.DAT file is placed on the Secondary node] *********************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Ensure the Primary node SSFS_HN1.KEY file is placed on the Secondary node] *********************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Ensure backup directory exists for HANA database with System Identifier HN1] *******************************************************************************
changed: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Check whether backup exists for SYSTEMDB database for System Identifier HN1] *******************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure backup exists for SYSTEMDB database for System Identifier HN1] **************************************************************************************
[DEPRECATION WARNING]: evaluating ('BACKUP_ID\n0 rows selected' in backup_exists_cmd_for_systemdb_result.stdout) | bool as a bare variable, this behaviour will go away and you might need 
to add |bool to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: evaluating ('BACKUP_ID\n0 rows selected' in backup_exists_cmd_for_systemdb_result.stdout) | bool as a bare variable, this behaviour will go away and you might need 
to add |bool to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
changed: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Check whether backup exists for tenant HN1 database for System Identifier HN1] *****************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure backup exists for tenant HN1 database for System Identifier HN1] ************************************************************************************
[DEPRECATION WARNING]: evaluating ('BACKUP_ID\n0 rows selected' in backup_exists_cmd_for_tenant_result.stdout) | bool as a bare variable, this behaviour will go away and you might need to
 add |bool to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: evaluating ('BACKUP_ID\n0 rows selected' in backup_exists_cmd_for_tenant_result.stdout) | bool as a bare variable, this behaviour will go away and you might need to
 add |bool to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
changed: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Check backup file exists for SYSTEMDB database for System Identifier HN1] **********************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Check backup file exists for tenant HN1 database for System Identifier HN1] ********************************************************************************
ok: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure Primary node is configured for System Replication] **************************************************************************************************
skipping: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Check replication state on primary] ************************************************************************************************************************
skipping: [10.1.1.5]
ok: [10.1.1.4]

TASK [hana-system-replication : Ensure options set in section [inifile_checker] of global.ini.] ********************************************************************************************
skipping: [10.1.1.5] => (item={'option': 'enable', 'value': 'true'}) 
skipping: [10.1.1.5] => (item={'option': 'interval', 'value': '3600'}) 
skipping: [10.1.1.5] => (item={'option': 'replication', 'value': 'true'}) 
changed: [10.1.1.4] => (item={'option': 'enable', 'value': 'true'})
changed: [10.1.1.4] => (item={'option': 'interval', 'value': '3600'})
changed: [10.1.1.4] => (item={'option': 'replication', 'value': 'true'})
[WARNING]: Module remote_tmp /usr/sap/HN1/home/.ansible/tmp did not exist and was created with a mode of 0700, this may cause issues when running as another user. To avoid this, create
the remote_tmp dir with the correct permissions manually


TASK [hana-system-replication : Determine if HANA is stopped] ******************************************************************************************************************************
skipping: [10.1.1.5]
ok: [10.1.1.4]

TASK [hana-system-replication : Ensure HANA is stopped] ************************************************************************************************************************************
skipping: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Check HANA is stopped] *************************************************************************************************************************************
skipping: [10.1.1.5]
ok: [10.1.1.4]

TASK [hana-system-replication : Determine if HANA is running] ******************************************************************************************************************************
skipping: [10.1.1.5]
ok: [10.1.1.4]

TASK [hana-system-replication : Ensure HANA is running] ************************************************************************************************************************************
skipping: [10.1.1.5]
changed: [10.1.1.4]

TASK [hana-system-replication : Check HANA is running] *************************************************************************************************************************************
skipping: [10.1.1.5]
ok: [10.1.1.4]

TASK [hana-system-replication : Determine if HANA is stopped] ******************************************************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure HANA is stopped] ************************************************************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Check HANA is stopped] *************************************************************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure Secondary node is registered as secondary in replication] *******************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Determine if HANA is running] ******************************************************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure HANA is running] ************************************************************************************************************************************
skipping: [10.1.1.4]
changed: [10.1.1.5]

TASK [hana-system-replication : Check HANA is running] *************************************************************************************************************************************
skipping: [10.1.1.4]
ok: [10.1.1.5]

TASK [hana-system-replication : Ensure replication status is active] ***********************************************************************************************************************
skipping: [10.1.1.5]
FAILED - RETRYING: Ensure replication status is active (30 retries left).
ok: [10.1.1.4]

PLAY [jumpboxes_linux] *********************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************
ok: [10.0.1.5]

TASK [linux-jumpboxes-install : Collect components for Linux jumpboxes] ********************************************************************************************************************
ok: [10.0.1.5] => (item={'authentication': {'type': 'key', 'username': 'rfallows'}, 'components': ['hana_studio_linux'], 'destroy_after_deploy': 'false', 'disk_type': 'StandardSSD_LRS', 'name': 'jumpbox-linux', 'os': {'offer': 'UbuntuServer', 'publisher': 'Canonical', 'sku': '16.04-LTS'}, 'private_ip_address': '10.0.1.5', 'public_ip_address': '13.72.100.34', 'size': 'Standard_D2s_v3'})
skipping: [10.0.1.5] => (item={'authentication': {'type': 'key', 'username': 'rfallows'}, 'components': ['ansible'], 'destroy_after_deploy': 'true', 'disk_type': 'StandardSSD_LRS', 'name': 'rti', 'os': {'offer': 'UbuntuServer', 'publisher': 'Canonical', 'sku': '18.04-LTS'}, 'private_ip_address': '10.0.1.6', 'public_ip_address': '13.72.98.242', 'size': 'Standard_D2s_v3'}) 

TASK [linux-jumpboxes-install : Flatten components to have one list for all Linux jumpboxes] ***********************************************************************************************
ok: [10.0.1.5]

TASK [Install components for Linux jumpboxes] **********************************************************************************************************************************************

TASK [components-install : Create dictionary with installation variant for HANA Studio] ****************************************************************************************************
skipping: [10.0.1.5] => (item={'component': 'hana_database', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_SERVER_LINUX_X86_64', 'installer': 'hdblcm'}]}) 
ok: [10.0.1.5] => (item={'component': 'hana_studio', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_STUDIO_LINUX_X86_64', 'installer': 'hdbinst'}, {'os': 'WINDOWS', 'media_loc': 'HDB_STUDIO_WINDOWS_X86_64', 'installer': 'hdbinst.exe'}]})
skipping: [10.0.1.5] => (item={'component': 'hana_client', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_CLIENT_LINUX_X86_64', 'installer': 'hdbinst'}, {'os': 'WINDOWS', 'media_loc': 'HDB_CLIENT_WINDOWS_X86_64', 'installer': 'hdbinst.exe'}]}) 
skipping: [10.0.1.5] => (item={'component': 'xs', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_SERVER_LINUX_X86_64', 'installer': 'hdblcm'}], 'depends_on': ['hana_database']}) 
skipping: [10.0.1.5] => (item={'component': 'shine', 'variants': [{'os': 'LINUX', 'media_loc': 'XSA_CONTENT_10', 'installer': 'xs'}], 'depends_on': ['hana_database', 'xs']}) 

TASK [components-install : Install HANA Studio on Linux] ***********************************************************************************************************************************
changed: [10.0.1.5] => (item={'os': 'LINUX', 'media_loc': 'HDB_STUDIO_LINUX_X86_64', 'installer': 'hdbinst'})
skipping: [10.0.1.5] => (item={'os': 'WINDOWS', 'media_loc': 'HDB_STUDIO_WINDOWS_X86_64', 'installer': 'hdbinst.exe'}) 

PLAY [jumpboxes_windows] *******************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************
ok: [10.0.1.4]

TASK [windows-jumpboxes-install : Collect components for Windows jumpboxes] ****************************************************************************************************************
ok: [10.0.1.4] => (item={'authentication': {'password': 'Passw0rd', 'type': 'password', 'username': 'rfallows'}, 'components': ['hana_studio_windows'], 'destroy_after_deploy': 'false', 'disk_type': 'StandardSSD_LRS', 'name': 'jumpbox-windows', 'os': {'offer': 'WindowsServer', 'publisher': 'MicrosoftWindowsServer', 'sku': '2016-Datacenter'}, 'private_ip_address': '10.0.1.4', 'public_ip_address': '40.87.17.76', 'size': 'Standard_D2s_v3'})

TASK [windows-jumpboxes-install : Flatten components to have one list for all Windows jumpboxes] *******************************************************************************************
ok: [10.0.1.4]

TASK [Install components for Windows jumpboxes] ********************************************************************************************************************************************

TASK [components-install : Create dictionary with installation variant for HANA Studio] ****************************************************************************************************
skipping: [10.0.1.4] => (item={'component': 'hana_database', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_SERVER_LINUX_X86_64', 'installer': 'hdblcm'}]}) 
ok: [10.0.1.4] => (item={'component': 'hana_studio', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_STUDIO_LINUX_X86_64', 'installer': 'hdbinst'}, {'os': 'WINDOWS', 'media_loc': 'HDB_STUDIO_WINDOWS_X86_64', 'installer': 'hdbinst.exe'}]})
skipping: [10.0.1.4] => (item={'component': 'hana_client', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_CLIENT_LINUX_X86_64', 'installer': 'hdbinst'}, {'os': 'WINDOWS', 'media_loc': 'HDB_CLIENT_WINDOWS_X86_64', 'installer': 'hdbinst.exe'}]}) 
skipping: [10.0.1.4] => (item={'component': 'xs', 'variants': [{'os': 'LINUX', 'media_loc': 'HDB_SERVER_LINUX_X86_64', 'installer': 'hdblcm'}], 'depends_on': ['hana_database']}) 
skipping: [10.0.1.4] => (item={'component': 'shine', 'variants': [{'os': 'LINUX', 'media_loc': 'XSA_CONTENT_10', 'installer': 'xs'}], 'depends_on': ['hana_database', 'xs']}) 

TASK [components-install : Copy HANA Studio Installer to windows jumpboxes] ****************************************************************************************************************
skipping: [10.0.1.4] => (item={'os': 'LINUX', 'media_loc': 'HDB_STUDIO_LINUX_X86_64', 'installer': 'hdbinst'}) 
changed: [10.0.1.4] => (item={'os': 'WINDOWS', 'media_loc': 'HDB_STUDIO_WINDOWS_X86_64', 'installer': 'hdbinst.exe'})

TASK [components-install : Install Visual C++ Redistributable Packages for Visual Studio 2013] *********************************************************************************************
changed: [10.0.1.4]

TASK [components-install : Install HANA Studio on Windows jumpboxes] ***********************************************************************************************************************
skipping: [10.0.1.4] => (item={'os': 'LINUX', 'media_loc': 'HDB_STUDIO_LINUX_X86_64', 'installer': 'hdbinst'}) 
changed: [10.0.1.4] => (item={'os': 'WINDOWS', 'media_loc': 'HDB_STUDIO_WINDOWS_X86_64', 'installer': 'hdbinst.exe'})

PLAY RECAP *********************************************************************************************************************************************************************************
10.0.1.4                   : ok=7    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
10.0.1.5                   : ok=11   changed=5    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
10.1.1.4                   : ok=77   changed=44   unreachable=0    failed=0    skipped=23   rescued=0    ignored=0   
10.1.1.5                   : ok=80   changed=47   unreachable=0    failed=0    skipped=20   rescued=0    ignored=0   
localhost                  : ok=16   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

</details>

### CODE REVIEW

File: deploy/v2/ansible/roles/hana-system-replication/tasks/set_global_ini_values.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+---
+
+    # HANA must be restarted after setting values in global.ini
+  - name: Ensure options set in section [inifile_checker] of global.ini.
```
Comment (rukawata) [inline deploy/v2/ansible/roles/hana-system-replication/tasks/set_global_ini_values.yml:4]:
```suggestion
     # Reference: 
     # - SAP Help document: Monitoring and Replicating INI File Parameter Changes (https://help.sap.com/viewer/6b94445c94ae495c83a19646e7c3fd56/2.0.04/en-US/f65a9c62f60f4d4bb91620dd959ee68d.html)
     # - SAP Note 2800357 infile_checker not able to Replicate Parameters although "replicate" is set to True
  - name: Ensure options set in section [inifile_checker] of global.ini.
```

File: deploy/v2/ansible/roles/hana-system-replication/tasks/set_global_ini_values.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      state: present
+      option: "{{ item.option }}"
+      value: "{{ item.value }}"
+    with_items:
+      - option: enable
+        value: "true"
+      - option: interval
+        value: "3600"
+      - option: replication
+        value: "true"
```
Comment (rukawata) [inline deploy/v2/ansible/roles/hana-system-replication/tasks/set_global_ini_values.yml:17]:
```suggestion
      - option: replicate
        value: "true"
      - option: fixed_exclusion_global.ini/SYSTEM
        value: "storage/*, persistence/*path*, internal_hostname_resolution/*, public_hostname_resolution/*, system_replication*, multidb/mode, system_landscape_hostname_virtualization/sldsystemhome"
      
```

File: deploy/v2/ansible/roles/hana-system-replication/vars/main.yml [hunk 3]
```diff
@@ -8,15 +8,15 @@
...
 
 sid_admin_user: "{{ sid | lower }}adm"
 
 hsr_dat_name: "SSFS_{{ sid | upper }}.DAT"
 hsr_key_name: "SSFS_{{ sid | upper }}.KEY"
 
-tmp_ssfs_dat: "{{ tmp_ssfs_dir }}/{{ hsr_dat_name }}"
-tmp_ssfs_key: "{{ tmp_ssfs_dir }}/{{ hsr_key_name }}"
+tmp_ssfs_dat: "{{ tmp_copy_dir }}/{{ hsr_dat_name }}"
+tmp_ssfs_key: "{{ tmp_copy_dir }}/{{ hsr_key_name }}"
```
Comment (rukawata) [inline deploy/v2/ansible/roles/hana-system-replication/vars/main.yml:19]: What's the reason of variable tmp_ssfs_dir naming change to `tmp_copy_dir`? I assume this is for a different backlog item where we change the way we copy DAT/KEY files over?

Comment (robfallows) [reply deploy/v2/ansible/roles/hana-system-replication/vars/main.yml:19]: Ah. Good catch. That was leakage from another branch I was working on. I'll revert that in this branch, as it's unnecessary here.

**Pull request merged**